pkg/logstructured/logstructured.go

```diff
@@ -274,81 +357,65 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 
 	go func() {
 		defer wg.Done()
+		var lastRev int64
 		rev, events, err := l.log.List(ctx, "/", "", 1000, 0, false)
 		for len(events) > 0 {
 			if err != nil {
-				logrus.Errorf("failed to read old events for ttl")
+				logrus.Errorf("Failed to read old events for ttl, %v", err)
 				return
 			}
 
 			for _, event := range events {
 				if event.KV.Lease > 0 {
 					result <- event
 				}
+
+				if event.KV.ModRevision > lastRev {
+					lastRev = event.KV.ModRevision
+				}
 			}
 
 			_, events, err = l.log.List(ctx, "/", events[len(events)-1].KV.Key, 1000, rev, false)
 		}
+		lastListRevision <- lastRev
 	}()
 
 	go func() {
 		defer wg.Done()
-		for events := range l.log.Watch(ctx, "/") {
+		revision := <-lastListRevision
```

Comment:
check if the channel read returned 0 - if so then the list failed and we should error out here too
```suggestion
		revision := <-lastListRevision
```

Reply:
>  Question: is there any possibility of a race condition here? What happens if the queue worker removes a kv from the map in the time between it is returned from loadTTLEventKV, and when we check to see if it is older than the current kv? I think that should be fine?

I have already noticed it. If use sync.map, there is a race. So I use mutex + map to avoid.